X86/X86MachineInstructionRaiser.cpp

```diff
@@ -2447,6 +2447,140 @@ bool X86MachineInstructionRaiser::raiseDivideInstr(const MachineInstr &MI,
+
+  unsigned Op2Index;
+  if (isMemBT) {
+    assert(MemRefValue != nullptr && "Null memory reference value encountered "
+                                     "while raising compare instruction");
+    assert(memoryRefOpIndex == 0 &&
+           "Unexpected memory operand index in compare instruction");
+
+    LLVMContext &Ctx(MF.getFunction().getContext());
+    auto MemOpSize = getInstructionMemOpSize(MI.getOpcode());
+    assert(MemOpSize != 0 && "Expected mem op size to be > 0");
+
+    Type *OpTy = Type::getIntNTy(Ctx, MemOpSize * 8);
+
+    auto LoadedVal =
+        loadMemoryRefValue(MI, MemRefValue, memoryRefOpIndex, OpTy);
+
+    OpValues[0] = LoadedVal;
+    Op2Index = X86::AddrNumOperands;
+  } else {
+    // The instruction operands do not reference memory
+    unsigned Op1Index;
+
+    if (NumImplicitUses == 1) {
+      // If an implicit operand is used, that is op1.
+      MCPhysReg UseReg = MCIDesc.ImplicitUses[0];
+      Op1Index = MI.findRegisterUseOperandIdx(UseReg, false, nullptr);
+      Op2Index = MCIDesc.getNumDefs() == 0 ? 0 : 1;
+    } else {
+      // Explicit operands are used
+      Op1Index = MCIDesc.getNumDefs() == 0 ? 0 : 1;
+      Op2Index = Op1Index + 1;
+    }
+
+    MachineOperand CmpOp1 = MI.getOperand(Op1Index);
+
+    assert((CmpOp1.isReg()) &&
+           "Unhandled first operand type in bt instruction");
+
+    OpValues[0] = getRegOrArgValue(CmpOp1.getReg(), MBBNo);
+  }
+
+  MachineOperand Op2 = MI.getOperand(Op2Index);
+  Value *Op2Val = nullptr;
+  if (Op2.isReg()) {
+    Op2Val = raisedValues->castValue(getRegOrArgValue(Op2.getReg(), MBBNo),
+                                     OpValues[0]->getType(), RaisedBB);
+  } else if (Op2.isImm()) {
+    Op2Val = ConstantInt::get(OpValues[0]->getType(), Op2.getImm());
+  } else {
+    LLVM_DEBUG(MI.dump());
+    assert(false && "Unhandled operand type in compare instruction");
+  }
+  OpValues[1] = Op2Val;
+
+  assert(OpValues[0] != nullptr && OpValues[1] != nullptr &&
+         "Unable to materialize compare operand values");
+
+  assert((OpValues[0]->getType() == OpValues[1]->getType()) &&
+         "Mis-matched operand types encountered while raising bt instruction");
+
+  auto BitWidth = OpValues[0]->getType()->getPrimitiveSizeInBits();
+  auto ModuloOperand = ConstantInt::get(OpValues[1]->getType(), BitWidth);
+  auto PositionModulo =
+      BinaryOperator::CreateURem(OpValues[1], ModuloOperand, "", RaisedBB);
+
+  auto ConstOne = ConstantInt::get(OpValues[0]->getType(), 1);
+  auto BitMask =
+      BinaryOperator::CreateShl(ConstOne, PositionModulo, "", RaisedBB);
+
+  auto BitValue = BinaryOperator::CreateAnd(OpValues[0], BitMask, "", RaisedBB);
+  auto ConstZero = ConstantInt::get(OpValues[0]->getType(), 0);
+  auto CFValue = ICmpInst::Create(Instruction::ICmp, ICmpInst::ICMP_NE,
+                                  BitValue, ConstZero, "", RaisedBB);
+
+  raisedValues->setEflagValue(EFLAGS::CF, MI.getParent()->getNumber(), CFValue);
+
```

Comment:
Please add comments to outline the different BinaryOperator instruction being created based on the opcode.

Reply:
I've added some comments